cpp/src/arrow/compute/expression.h

```diff
@@ -0,0 +1,261 @@
+SIMPLE_EXPR_FACTORY(boolean);
+SIMPLE_EXPR_FACTORY(int8);
+SIMPLE_EXPR_FACTORY(int16);
+SIMPLE_EXPR_FACTORY(int32);
+SIMPLE_EXPR_FACTORY(int64);
+SIMPLE_EXPR_FACTORY(uint8);
+SIMPLE_EXPR_FACTORY(uint16);
+SIMPLE_EXPR_FACTORY(uint32);
+SIMPLE_EXPR_FACTORY(uint64);
+SIMPLE_EXPR_FACTORY(half_float);
+SIMPLE_EXPR_FACTORY(float_);
+SIMPLE_EXPR_FACTORY(double_);
+SIMPLE_EXPR_FACTORY(binary);
+SIMPLE_EXPR_FACTORY(utf8);
+
+class ARROW_EXPORT List : public ScalarExpr, public value::List {
+ public:
+  List(std::shared_ptr<const Operation> op, std::shared_ptr<LogicalType> type);
+  using ScalarExpr::kind;
+};
+
+class ARROW_EXPORT Struct : public ScalarExpr, public value::Struct {
+ public:
+  Struct(std::shared_ptr<const Operation> op, std::shared_ptr<LogicalType> type);
+  using ScalarExpr::kind;
+};
+
+}  // namespace scalar
+
+namespace array {
+
+#define DECLARE_ARRAY_EXPR(TYPE)                                   \
+  class ARROW_EXPORT TYPE : public ArrayExpr, public value::TYPE { \
+   public:                                                         \
+    explicit TYPE(std::shared_ptr<const Operation> op);            \
+    using ArrayExpr::kind;                                         \
+  };
+
+DECLARE_ARRAY_EXPR(Null)
+DECLARE_ARRAY_EXPR(Bool)
+DECLARE_ARRAY_EXPR(Int8)
+DECLARE_ARRAY_EXPR(Int16)
+DECLARE_ARRAY_EXPR(Int32)
+DECLARE_ARRAY_EXPR(Int64)
+DECLARE_ARRAY_EXPR(UInt8)
+DECLARE_ARRAY_EXPR(UInt16)
+DECLARE_ARRAY_EXPR(UInt32)
+DECLARE_ARRAY_EXPR(UInt64)
+DECLARE_ARRAY_EXPR(HalfFloat)
+DECLARE_ARRAY_EXPR(Float)
+DECLARE_ARRAY_EXPR(Double)
+DECLARE_ARRAY_EXPR(Binary)
+DECLARE_ARRAY_EXPR(Utf8)
+
+SIMPLE_EXPR_FACTORY(null);
+SIMPLE_EXPR_FACTORY(boolean);
+SIMPLE_EXPR_FACTORY(int8);
+SIMPLE_EXPR_FACTORY(int16);
+SIMPLE_EXPR_FACTORY(int32);
+SIMPLE_EXPR_FACTORY(int64);
+SIMPLE_EXPR_FACTORY(uint8);
+SIMPLE_EXPR_FACTORY(uint16);
+SIMPLE_EXPR_FACTORY(uint32);
+SIMPLE_EXPR_FACTORY(uint64);
+SIMPLE_EXPR_FACTORY(half_float);
+SIMPLE_EXPR_FACTORY(float_);
+SIMPLE_EXPR_FACTORY(double_);
+SIMPLE_EXPR_FACTORY(binary);
+SIMPLE_EXPR_FACTORY(utf8);
+
+class ARROW_EXPORT List : public ArrayExpr, public value::List {
+ public:
+  List(std::shared_ptr<const Operation> op, std::shared_ptr<LogicalType> type);
+  using ArrayExpr::kind;
+};
+
+class ARROW_EXPORT Struct : public ArrayExpr, public value::Struct {
+ public:
+  Struct(std::shared_ptr<const Operation> op, std::shared_ptr<LogicalType> type);
+  using ArrayExpr::kind;
+};
+
+}  // namespace array
+
+#undef SIMPLE_EXPR_FACTORY
+
+template <typename T, typename ObjectType>
+inline bool InheritsFrom(const ObjectType* obj) {
+  return dynamic_cast<const T*>(obj) != NULLPTR;
+}
+
+template <typename T, typename ObjectType>
+inline bool InheritsFrom(const ObjectType& obj) {
+  return dynamic_cast<const T*>(&obj) != NULLPTR;
+}
+
+/// \brief Construct a ScalarExpr containing an Operation given a logical type
+ARROW_EXPORT
+Status GetScalarExpr(std::shared_ptr<const Operation> op, std::shared_ptr<LogicalType> ty,
```

Comment:
Aside: we should look into creating a StatusOr class (https://github.com/protocolbuffers/protobuf/blob/master/src/google/protobuf/stubs/statusor.h)

Reply:
https://jira.apache.org/jira/browse/ARROW-4800